lib/src/value/color.ts

```diff
@@ -3,202 +3,372 @@
+ColorSpace.register(OKLab);
+ColorSpace.register(P3);
+ColorSpace.register(ProPhoto);
+ColorSpace.register(REC_2020);
+ColorSpace.register(XYZ_D50);
+ColorSpace.register(XYZ_D65);
+ColorSpace.register(sRGB);
+ColorSpace.register(sRGB_Linear);
+
+/** The HSL color space name. */
+type ColorSpaceHsl = 'hsl';
+
+/** The HSL color space channel names. */
+type ChannelNameHsl = 'hue' | 'saturation' | 'lightness' | 'alpha';
+
+/** The HWB color space name. */
+type ColorSpaceHwb = 'hwb';
+
+/** The HWB color space channel names. */
+type ChannelNameHwb = 'hue' | 'whiteness' | 'blackness' | 'alpha';
+
+/** The Lab / Oklab color space names. */
+type ColorSpaceLab = 'lab' | 'oklab';
+
+/** The Lab / Oklab color space channel names. */
+type ChannelNameLab = 'lightness' | 'a' | 'b' | 'alpha';
+
+/** The LCH / Oklch color space names. */
+type ColorSpaceLch = 'lch' | 'oklch';
+
+/** The LCH / Oklch color space channel names. */
+type ChannelNameLch = 'lightness' | 'chroma' | 'hue' | 'alpha';
+
+/** Names of color spaces with RGB channels. */
+type ColorSpaceRgb =
+  | 'a98-rgb'
+  | 'display-p3'
+  | 'prophoto-rgb'
+  | 'rec2020'
+  | 'rgb'
+  | 'srgb'
+  | 'srgb-linear';
+
+/** RGB channel names. */
+type ChannelNameRgb = 'red' | 'green' | 'blue' | 'alpha';
+
+/** Names of color spaces with XYZ channels. */
+type ColorSpaceXyz = 'xyz' | 'xyz-d50' | 'xyz-d65';
+
+/** XYZ channel names. */
+type ChannelNameXyz = 'x' | 'y' | 'z' | 'alpha';
+
+/** All supported color space channel names. */
+type ChannelName =
+  | ChannelNameHsl
+  | ChannelNameHwb
+  | ChannelNameLab
+  | ChannelNameLch
+  | ChannelNameRgb
+  | ChannelNameXyz;
+
+/** All supported color space names. */
+export type KnownColorSpace =
+  | ColorSpaceHsl
+  | ColorSpaceHwb
+  | ColorSpaceLab
+  | ColorSpaceLch
+  | ColorSpaceRgb
+  | ColorSpaceXyz;
+
+type Channels = {
+  [key in ChannelName]?: number | null;
+};
+
+function getColorSpace(options: Channels) {
+  if (typeof options.red === 'number') {
+    return 'rgb';
+  }
+  if (typeof options.saturation === 'number') {
+    return 'hsl';
+  }
+  if (typeof options.whiteness === 'number') {
+    return 'hwb';
+  }
+  throw valueError('No color space found');
 }
 
-interface HwbColor {
-  hue: number;
-  whiteness: number;
-  blackness: number;
-  alpha?: number;
+function emitColor4ApiDeprecation(name: string) {
+  console.warn(`\`${name}\` is deprecated, use \`channel\` instead.`);
 }
 
 /** A SassScript color. */
 export class SassColor extends Value {
-  private redInternal?: number;
-  private greenInternal?: number;
-  private blueInternal?: number;
-  private hueInternal?: number;
-  private saturationInternal?: number;
-  private lightnessInternal?: number;
-  private readonly alphaInternal: number;
-
-  constructor(color: RgbColor);
-  constructor(color: HslColor);
-  constructor(color: HwbColor);
-  constructor(color: RgbColor | HslColor | HwbColor) {
+  private color: PlainColorObject;
+  private format: 'rgb' | null = null;
```

Comment:
I don't understand this field—how does this differ from `color.space`? When is it null?

Reply:
Yeah, this was a first draft and at least needs renaming/documentation, if it doesn't change entirely. Basically ColorJS always treats `rgb` as a format of `srgb`, not its own color space. So any usage of `rgb` here in embedded-host will be handled internally as an `srgb` color, and will require manual mapping to scale 0-255 down to 0-1 (or vice versa).

Reply:
Adjusted in https://github.com/sass/embedded-host-node/pull/259/commits/a27e7b6b50d4082f47de0a4a3bb2a520026ddefc